### ISSUE TITLE
IoU reports incorrect performance on pretrained PyTorch segmentation model

### ISSUE THREAD

**Issue by aretor** [PyTorchLightning/metrics#552]
## 🐛 Bug

When evaluating a pretrained segmentation model (PyTorch `deeplabv3+_resnet101`) on a dataset (PyTorch `VOCSegmentation`), with `torchmetrics.IoU`, the CLI reporter returns ~34%, however when printing `iou.compute()` returns `~68%`. No training of the model has been performed, just a simple evaluation with `trainer.validate`.

### To Reproduce

1. Use the code sample below
1. Run `python train_script --max_epochs 1 --seed 314 --gpus 1 --log_every_n_steps 1 --lr 0.006 --batch_size 12 --log_gpu_memory all --loader_workers 40 --persistent_loader_workers`
1. Check accuracy after the validation

#### Code sample

```python
from argparse import ArgumentParser
import os.path as osp

import torch
import torch.nn as nn
import torch.optim as optim
from torchvision.models.segmentation import deeplabv3_resnet101
from torchvision.datasets import VOCSegmentation

from torchmetrics import IoU, Accuracy, MetricCollection
from pytorch_lightning.utilities.seed import seed_everything
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DDPPlugin
from pytorch_lightning.loggers import TensorBoardLogger


from typing import Optional, Union, List, Dict
import os.path as osp
from argparse import ArgumentParser
from filelock import FileLock

from torch.utils.data import DataLoader

from pytorch_lightning import LightningDataModule


from torchvision.transforms import functional as F
import numpy as np


class ImageToTensor(object):
    def __call__(self, image):
        image = F.resize(image, (512, 512))
        return F.to_tensor(image)


class TargetToTensor(object):
    def __call__(self, target):
        target = F.resize(target, (512, 512))
        target = torch.as_tensor(np.array(target),
                                 dtype=torch.float64).unsqueeze(0)
        target[target == 255.] = 21.
        return target

class VOCSegmentationAdvanced(VOCSegmentation):
    n_classes = 21
    ignore_index = 21
    def __getitem__(self, index):
        img, target = super(VOCSegmentationAdvanced, self).__getitem__(index)
        id_ = osp.splitext(osp.split(self.images[index])[-1])[0]

        return {"img": img, "target": target, "id": id_}


def lightning_parser(parser=None):
    parser = parser or ArgumentParser()
    parser.add_argument("--seed", type=int, default=None,
        help="The seed for reproducibility."
    )
    parser.add_argument("--exp_dir", type=str,
        default=osp.join(".", "results", "lightning"),
        help="The folder where to store the experiment."
    )
    parser.add_argument("--logger_dir", type=str,
        default="logs",
        help="The folder where to store the TensorBoard logs. The logs are "
             "stored in <exp_dir>/<logger_dir>.")
    parser.add_argument("--checkpoint_dir", type=str,
        default="checkpoints",
        help="The folder where to store the checkponts. The checkpoints are "
             "stored in <exp_dir>/<checkpoint_dir>.")

    DeepLabModule.setup_arg_parser(parser)
    VOCSegmentationDataModule.add_argparse_args(parser)
    Trainer.add_argparse_args(parser)

    return parser


class VOCSegmentationDataModule(LightningDataModule):
    def __init__(self, dset_name, root, batch_size=4, loader_workers=0,
                 persistent_loader_workers=True, download_data=False, *args,
                 **kwargs):
        super().__init__()
        self.root = root
        self.dset_name = dset_name
        self.batch_size = batch_size or 1
        self.loader_workers = loader_workers
        self.persistent_loader_workers = persistent_loader_workers
        self.download_data = download_data

    @classmethod
    def add_argparse_args(cls, parent_parser: ArgumentParser, **kwargs) -> ArgumentParser:
        parser = parent_parser.add_argument_group("data arguments")
        parser.add_argument("--dataset", dest="dset_name", type=str,
                            default="pascalvoc")
        parser.add_argument("--root", type=str,
                            default=osp.join(".", "datasets", "datasets"))
        parser.add_argument("--batch_size", type=int, default=None)
        parser.add_argument("--loader_workers", type=int, default=0)
        parser.add_argument("--persistent_loader_workers", action="store_true")
        parser.add_argument("--download_data", action="store_true")
        return super().add_argparse_args(parent_parser, **kwargs)

    def prepare_data(self) -> None:
        for split in ["train", "val"]:
            with FileLock(osp.expanduser(osp.join("~", "data.lock"))):
                 VOCSegmentationAdvanced(root=self.root,
                                         transform=ImageToTensor(),
                                         target_transform=TargetToTensor(),
                                         year="2012",
                                         image_set=split,
                                         download=not osp.exists(self.root))

    def setup(self, stage: Optional[str]) -> None:
        download = not osp.exists(self.root)
        self.train_dataset = VOCSegmentationAdvanced(root=self.root,
                                                     transform=ImageToTensor(),
                                                     target_transform=TargetToTensor(),
                                                     year="2012",
                                                     image_set="train",
                                                     download=download)
        self.val_dataset = VOCSegmentationAdvanced(root=self.root,
                                                   transform=ImageToTensor(),
                                                   target_transform=TargetToTensor(),
                                                   year="2012",
                                                   image_set="val",
                                                   download=download)


    def train_dataloader(
        self,
    ) -> Union[DataLoader, List[DataLoader], Dict[str, DataLoader]]:
        return DataLoader(
            self.train_dataset,
            batch_size=self.batch_size,
            num_workers=self.loader_workers,
            persistent_workers=self.persistent_loader_workers,
            pin_memory=True
        )

    def val_dataloader(
        self,
    ) -> Union[DataLoader, List[DataLoader], Dict[str, DataLoader]]:
        return DataLoader(
            self.val_dataset,
            batch_size=self.batch_size,
            num_workers=self.loader_workers,
            persistent_workers=self.persistent_loader_workers,
            pin_memory=True
        )



class DeepLabModule(LightningModule):
    @staticmethod
    def setup_arg_parser(parent_parser):
        parser = parent_parser.add_argument_group("model hyperparameters")
        parser.add_argument("--lr", "--learning_rate", type=float,
                            default=1e-8)
        parser.add_argument("--weight_decay", type=float, default=0.0)
        parser.add_argument("--momentum", type=float, default=0.9)
        return parser

    def __init__(self, config, init_dset, *args, **kwargs) -> None:
        super().__init__()

        self.learning_rate = config["lr"]
        self.weight_decay = config["weight_decay"]
        self.momentum = config["momentum"]

        num_classes = init_dset.n_classes
        ignore_index = init_dset.ignore_index

        self.model = deeplabv3_resnet101(pretrained=True)
        self.criterion = nn.CrossEntropyLoss(ignore_index=ignore_index)
        metrics = MetricCollection([
            Accuracy(num_classes=num_classes + 1, ignore_index=ignore_index),
            IoU(num_classes=num_classes + 1, ignore_index=ignore_index)
        ])

        self.val_metrics = metrics.clone(prefix="val_")    

        self.save_hyperparameters()

    def configure_optimizers(self):
        optimizer = optim.SGD(
            self.model.parameters(),
            lr=self.learning_rate,
            weight_decay=self.weight_decay,
            momentum=self.momentum,
        )
        return optimizer

    def forward(self, X):
        return self.model(X)["out"]

    def validation_step(self, batch, batch_index, *args, **kwargs):
        X, Y = batch["img"], batch["target"].squeeze(1).long()

        with torch.no_grad():
            logits = self(X)
            val_loss = 10.  # placeholder
            output = self.val_metrics(logits.argmax(dim=1), Y)
            self.log_dict({"val_loss": val_loss, **output}, on_step=True,
                          on_epoch=True, sync_dist=False)
            return val_loss
    

def get_data_args(args):
    return {
        "dset_name": args.dset_name,
        "root": args.root,
        "batch_size": args.batch_size,
        "loader_workers": args.loader_workers,
        "persistent_loader_workers": args.persistent_loader_workers,
        "download_data": args.download_data,
    }


def get_trainer_args(args):
    return {"max_epochs": args.max_epochs,
            "gpus": args.gpus,
            "log_every_n_steps": args.log_every_n_steps,
            "log_gpu_memory": args.log_gpu_memory,
            "auto_scale_batch_size": "binsearch" if args.batch_size is None \
                                     else False,
            }


def my_train(config, data_args, trainer_args, logger_dir):
    datamodule = VOCSegmentationDataModule(**data_args)
    datamodule.prepare_data()
    datamodule.setup(None)

    model = DeepLabModule(config, init_dset=datamodule.train_dataset)

    tb_logger = TensorBoardLogger(logger_dir)

    trainer = Trainer(**trainer_args, logger=tb_logger,
                      plugins=DDPPlugin(find_unused_parameters=False),
                      accelerator="ddp")

    trainer.validate(model, datamodule=datamodule)
    exit()


def main(args):
    seed_everything(args.seed, workers=True)

    data_args = get_data_args(args)
    trainer_args = get_trainer_args(args)
    config = {"lr": args.lr,
              "weight_decay": args.weight_decay,
              "momentum": args.momentum
              }
    logger_dir = osp.join(args.exp_dir, args.logger_dir)
    my_train(config, data_args, trainer_args, logger_dir)


if __name__ == "__main__":
    parser = lightning_parser()
    args = parser.parse_args()
    main(args)

```

### Expected behavior

The CLI reporter should report ~68%, instead of ~34%

### Environment

- PyTorch Version (e.g., 1.0): 1.8.1
- OS (e.g., Linux): Ubuntu 18.04
- How you installed PyTorch (`conda`, `pip`, source): pip (on conda environment)
- Python version: 3.8.8
- CUDA/cuDNN version: 10.1/...
- GPU models and configuration: Quadro  RTX 8000 x 2, Quadro RTX 5000 x 6
- Any other relevant information: ...

**Comment by SkafteNicki** [comment]
To me it seems like the logging is wrong
```python
    def validation_step(self, batch, batch_index, *args, **kwargs):
        X, Y = batch["img"], batch["target"].squeeze(1).long()

        with torch.no_grad():
            logits = self(X)
            val_loss = 10.  # placeholder
            output = self.val_metrics(logits.argmax(dim=1), Y)
            self.log_dict({"val_loss": val_loss, **output}, on_step=True,
                          on_epoch=True, sync_dist=False)
            return val_loss
```
since you are passing the calculated metric values (and not the metric objects themself) and using `on_epoch=True` you are essentially not calculating the global IoU metric but instead the average of all the individual batches. Without having tested it, I recommend changing your code to something like:
```python
    def validation_step(self, batch, batch_index, *args, **kwargs):
        X, Y = batch["img"], batch["target"].squeeze(1).long()

        with torch.no_grad():
            logits = self(X)
            val_loss = 10.  # placeholder
            self.val_metrics.update(logits.argmax(dim=1), Y)
            self.log_dict({"val_loss": val_loss}, on_step=True, on_epoch=True)
            return val_loss

    def validation_epoch_end(self, outputs):
        output = self.val_metrics.compute()
        self.log_dict(output)
```

**Comment by aretor** [comment]
I was trying to follow [this documentation](https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html#logging-torchmetrics), however, it seems that you cannot pass a `MetricCollection` object to `self.log_dict`. It returns the following error:

```
Exception has occurred: ValueError

`self.log(val_metrics, MetricCollection(
  (Accuracy): Accuracy()
  (IoU): IoU(),
  prefix=val_
)) ` was called, but `MetricCollection` values cannot be logged
```

I tried the following:

```python
    def validation_step(self, batch, batch_index, *args, **kwargs):
        X, Y = batch["img"], batch["target"].squeeze(1).long()

        with torch.no_grad():
            logits = self(X)
            val_loss = 10.  # placeholder
            self.val_metrics(logits.argmax(dim=1), Y)
            self.log_dict(
                {
                    "val_loss": val_loss,
                    "val_metrics": dict(self.val_metrics.named_children())
                },
                on_step=False, on_epoch=True, sync_dist=False)
            return val_loss
```
You can then access the metrics with `val_metrics_epoch_IoU` and the like. This works when calling validate only, but does not when training the model. In fact, it throws the following error:

```
  Exception has occurred: MisconfigurationException
  You called `self.log(val_metrics, ...)` twice in `validation_step` with different arguments. This is not allowed
```

I managed to solve both problems with the following workaround:

```python
    def validation_step(self, batch, batch_index, *args, **kwargs):
        X, Y = batch["img"], batch["target"].squeeze(1).long()

        with torch.no_grad():
            logits = self(X)
            val_loss = 10.  # placeholder
            self.val_metrics(logits.argmax(dim=1), Y)
            metric_logs = {self.val_metrics.prefix + k: v for k, v in dict(self.val_metrics.named_children()).items()}
            self.log_dict(
                {
                    "val_loss": val_loss,
                    **metric_logs
                },
                on_step=False, on_epoch=True, sync_dist=False)
            return val_loss
```

If there is some neater solution, please, point it out. If not, I think that the `log` method could be improved to process CollectionMetric correctly.
